### PR TITLE
feat(Workspaces) : add user to a workspace

### PR DESCRIPTION
**Pull request by cheikhgwane**
This PR is for adding a user to a workspace.

## Changes

Please list / describe the changes in the codebase for the reviewer(s).

- Add user to a workspace
- Send added user the workspace link via mail

## How/what to test

A few words about what needs to be tested, along with specific testing instructions if needed.

## Screenshots / screencast

If relevant, please add some screenshots or a short video.

### CODE REVIEW

File: hexa/workspaces/models.py [hunk 1]
```diff
@@ -43,18 +47,15 @@ class WorkspaceQuerySet(BaseQuerySet):
...
+        return self._filter_for_user_and_query_object(
+            user, Q(workspacemembership__user=user)
+        )
 
 
 class Workspace(Base):
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
     name = models.TextField()
     description = models.TextField(blank=True)
-    members = models.ManyToManyField(
```
Comment (qgerome) [inline hexa/workspaces/models.py:53]: Why did you remove this ?

File: hexa/workspaces/models.py [hunk 2]
```diff
@@ -82,12 +83,49 @@ def delete_if_has_perm(self, *, principal: User):
...
 
 
+class WorkspaceMembershipQuerySet(BaseQuerySet):
+    def filter_for_user(
+        self, user: typing.Union[AnonymousUser, User]
+    ) -> models.QuerySet:
+
+        return (
+            self.all()
+            if WorkspaceMembership.objects.filter(
```
Comment (qgerome) [inline hexa/workspaces/models.py:93]: I think you need to add a `.exists()` or `.first()` at the end to this no ? Here I think you just check if it returns a query set

File: hexa/workspaces/schema.py [hunk 3]
```diff
@@ -109,4 +123,51 @@ def resolve_delete_workspace(_, info, **kwargs):
         return {"success": False, "errors": ["PERMISSION_DENIED"]}
 
 
+@worskspace_mutations.field("createWorkspaceMember")
```
Comment (qgerome) [inline hexa/workspaces/schema.py:126]: I'd prefer "inviteWorkspaceMember" WDYT ?

File: hexa/workspaces/schema.py [hunk 4]
```diff
@@ -109,4 +123,51 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+        user = (
+            User.objects.get(email=input["userEmail"]) if "userEmail" in input else None
+        )
+
+        workspace_membership = WorkspaceMembership.objects.create_if_has_perm(
+            principal=request.user, workspace=workspace, user=user, role=input["role"]
+        )
+
+        send_mail(
+            title="You've been added to a Workspace",
```
Comment (qgerome) [inline hexa/workspaces/schema.py:143]: lazy_gettext

File: hexa/workspaces/schema.py [hunk 5]
```diff
@@ -109,4 +123,51 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+            template_name="workspaces/mails/new_member",
```
Comment (qgerome) [inline hexa/workspaces/schema.py:144]: invite_member

File: hexa/workspaces/templates/workspaces/mails/new_member.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+Your Have been added to an OpenHexa Workspace
```
Comment (qgerome) [inline hexa/workspaces/templates/workspaces/mails/new_member.txt:1]: Would be nice to display the name of the workspace and the name of the user that sent the invitation.

File: hexa/workspaces/templates/workspaces/mails/new_member.html [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{% load i18n %}
+{% autoescape off %}
+{% blocktranslate %}Your have been added to an OpenHexa Workspace.{% endblocktranslate %}
+
+{% translate "Please go to the following page to access it:" %}
+
+{{ url }}
```
Comment (qgerome) [inline hexa/workspaces/templates/workspaces/mails/new_member.html:7]: You can render this as a link at least.

**Review by qgerome — CHANGES_REQUESTED** [review]

File: hexa/workspaces/models.py (hunk 2 above)
Comment (cheikhgwane) [reply hexa/workspaces/models.py:93]: Yes exactly. I'll add it

File: hexa/workspaces/templates/workspaces/mails/new_member.txt (hunk 6 above)
Comment (cheikhgwane) [reply hexa/workspaces/templates/workspaces/mails/new_member.txt:1]: Yes, on it.

File: hexa/workspaces/templates/workspaces/mails/new_member.html (hunk 7 above)
Comment (cheikhgwane) [reply hexa/workspaces/templates/workspaces/mails/new_member.html:7]: Yes.

File: hexa/workspaces/schema.py (hunk 3 above)
Comment (cheikhgwane) [reply hexa/workspaces/schema.py:129]: Yes, it's better

File: hexa/workspaces/models.py (hunk 1 above)
Comment (qgerome) [reply hexa/workspaces/models.py:53]: I think this is exactly like the `Team` & `Membership` models. 
The relation is a M2M so I think it's better represented by a `ManytoManyField`

@pvanliefland Do you have an opinion on this ?

File: hexa/workspaces/schema.py [hunk 8]
```diff
@@ -18,10 +22,23 @@
 worskspace_mutations = MutationType()
```
Comment (qgerome) [inline hexa/workspaces/schema.py:22]: There is a typo in the name

File: hexa/workspaces/schema.py [hunk 9]
```diff
@@ -109,4 +126,54 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+@worskspace_mutations.field("inviteWorkspaceMember")
+def resolve_create_workspace_member(_, info, **kwargs):
+    request: HttpRequest = info.context["request"]
+    input = kwargs["input"]
+    try:
+        workspace: Workspace = Workspace.objects.filter_for_user(request.user).get(
+            id=input["workspaceId"]
+        )
+        user = (
+            User.objects.get(email=input["userEmail"]) if "userEmail" in input else None
```
Comment (qgerome) [inline hexa/workspaces/schema.py:138]: userEmail is always present in input

File: hexa/workspaces/schema.py [hunk 10]
```diff
@@ -109,4 +126,54 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+        )
+
+        workspace_membership = WorkspaceMembership.objects.create_if_has_perm(
+            principal=request.user, workspace=workspace, user=user, role=input["role"]
+        )
+        send_mail(
+            title=gettext_lazy("You've been added to a Workspace"),
```
Comment (qgerome) [inline hexa/workspaces/schema.py:145]: "You've been added to the workspace '{name}'"

File: hexa/workspaces/schema.py [hunk 11]
```diff
@@ -109,4 +126,54 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+            template_name="workspaces/mails/invite_member",
+            template_variables={
+                "workspace": workspace.name,
+                "owner": "{} {}".format(
+                    request.user.first_name, request.user.last_name
```
Comment (qgerome) [inline hexa/workspaces/schema.py:150]: request.user.display_name

File: hexa/workspaces/schema.py [hunk 12]
```diff
@@ -109,4 +126,54 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+                ),
+                "url": "{url}/workspaces/{workspace_id}".format(
```
Comment (qgerome) [inline hexa/workspaces/schema.py:152]: url -> workspace_url

**Review by qgerome — CHANGES_REQUESTED** [review]
Could you add tests where you get the members of a workspace ?

File: hexa/workspaces/schema.py (hunk 9 above)
Comment (cheikhgwane) [reply hexa/workspaces/schema.py:138]: Right !

File: hexa/workspaces/models.py (hunk 1 above)
Comment (pvanliefland) [reply hexa/workspaces/models.py:53]: You are both right ;) 

Ideally you should use a many-to-many field with an intermediate model like explain here: https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.ManyToManyField.through.

This is indeed what we did for `Team` and `Membership`

Comment (cheikhgwane) [reply hexa/workspaces/models.py:53]: Okay. I'll revert it.

File: hexa/workspaces/schema.py [hunk 13]
```diff
@@ -109,4 +155,55 @@ def resolve_delete_workspace(_, info, **kwargs):
...
+                ),
+            },
+            recipient_list=[user.email],
+        )
+        return {
+            "success": True,
+            "errors": [],
+            "workspace_membership": workspace_membership,
+        }
+    except (Workspace.DoesNotExist, User.DoesNotExist):
```
Comment (qgerome) [inline hexa/workspaces/schema.py:188]: You have to split the two cases since it's not the same to not find the user as to not find the workspace you try to use.

File: hexa/workspaces/templates/workspaces/mails/invite_member.html [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{% load i18n %}
+{% autoescape off %}
+{% blocktranslate %}Your have been added to the workspace {{workspace}} by {{owner}}.{% endblocktranslate %}
```
Comment (qgerome) [inline hexa/workspaces/templates/workspaces/mails/invite_member.html:3]:
```suggestion
{% blocktranslate %}Your have been added to the workspace <b>{{workspace}}</b> by {{owner}}.{% endblocktranslate %}
```

**Review by qgerome — CHANGES_REQUESTED** [review]

**Comment by qgerome** [comment]
The interpolation of the subject of the mail is not correct
CleanShot 2023-01-13 at 12 12 03@2x
